[CDAP-15738] Upgrades node version and corresponding node modules for compatibility

ajainarayanan opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-15738
Build: https://builds.cask.co/browse/CDAP-UDUT377-1

- Upgrades node version to 10.16.2
- Upgrades babel and gulp node modules for compatibility
- Fixes Gulpfile and webpack build config files based on upgrade

yukiej commented on cdap-ui/server.js:
```diff
@@ -112,4 +111,21 @@ cdapConfigurator.getCDAPConfig()
+        allowedOrigin = `${nodejsprotocol}://${nodejsserver}`;
+      }
+      if (req.headers.origin !== allowedOrigin) {
+        log.info('Unknown Orign: ' + req.headers.origin);
```
nit: typo in "origin"

yukiej reviewed (commented):
lgtm!

elfenheart commented:
The connection seemed to get rejected on Sandbox as well (probably due to localhost vs 127.0.0.1). However, even when the connection is rejected, I can still use the UI and do queries. Is that expected?

ajainarayanan commented:
Talked to @elfenheart about it. Since we are using sockjs it resorts to use other modes of communication to polyfil websockets (xhr_polling). That is expected.

elfenheart commented on cdap-ui/server.js:
```diff
@@ -112,4 +111,21 @@ cdapConfigurator.getCDAPConfig()
+      const nodejsprotocol = (cdapConfig['ssl.external.enabled'] === 'true') ? 'https' : 'http';
+      let allowedOrigin;
+      if (nodejsport) {
+        allowedOrigin = `${nodejsprotocol}://${nodejsserver}:${nodejsport}`;
```
`127.0.0.1` and `localhost` should be interchangeable. Currently cdap-default for sandbox is bind to `127.0.0.1`

ajainarayanan commented:
Currently on-hold as checking for origin alone will be problematic in some environments and is not the full solution. Working on adding a csrf-like token that will be validated on websocket upgrade or any http request.

ajainarayanan committed:
[CDAP-15738] Upgrades NodeJS to version 10.16.2

  - Updrades all relevant node modules
  - Updates all jest related modules
  - Updates build process based on upgraded node modules
  - Fixes logs4js json file to be adhere to new format
  - Updates pom.xml and build scripts to use node 10.16.2+
  - Modifies cdap-distribution build scripts to use Nodejs version 10.16.2
  - Updates cdap-docs to have the NODE_JS_MIN_VERSION as 10.16.2
  - Updates nodejs binary checksum in distributed build files

ajainarayanan commented:
Changes are up to date. Fixing tests to pass.

elfenheart commented on cdap-ui/package.json:
```diff
@@ -38,14 +38,20 @@
+    "@babel/plugin-proposal-object-rest-spread": "7.0.0",
+    "@babel/plugin-proposal-throw-expressions": "^7.0.0",
+    "@babel/plugin-syntax-dynamic-import": "^7.0.0",
+    "@babel/plugin-syntax-import-meta": "^7.0.0",
```
nit: remove the `^` and redo yarn install

elfenheart reviewed (commented):
I think you'll need to update pom.xml to change the node version as well

yukiej commented:
Should the cdap-ui/BUILD.rst get updated as well, with the nodeJS version? https://github.com/cdapio/cdap/blob/develop/cdap-ui/BUILD.rst

yukiej reviewed (approved):
One question/comment, otherwise LGTM!

dereklwood commented on cdap-distributions/src/packer/files/cdap-sdk.json:
```diff
@@ -4,7 +4,7 @@
+    "version": "10.16.2",
     "binary": {
       "checksum": {
         "linux_x64": "115c7bd133170fd7a1bf408b2e293021e4b5a80a66a4962829ce5d362ce43762"
```
this needs updating

elfenheart reviewed (approved):
👍 LGTM

ajainarayanan commented:
Three consecutive green build: [Build 1](https://builds.cask.co/browse/CDAP-UDUT377-13), [Build 2](https://builds.cask.co/build/result/gotoBuildResult.action?buildKey=CDAP-UDUT377&buildNumber=12&returnUrl=%2Fbrowse%2FCDAP-UDUT377-13), [Build 3](https://builds.cask.co/build/result/gotoBuildResult.action?buildKey=CDAP-UDUT377&buildNumber=11&returnUrl=%2Fbrowse%2FCDAP-UDUT377-13)
